### DISCUSSION TITLE
Is there a way to write in VS Code and Preview in the Browser?

### DISCUSSION THREAD

**Discussion by tonyhschu** [elalish/manifold#727] [Q&A] [answered]
I've been using JSCAD and JSCAD now (See: https://github.com/danmarshall/jscad-now) as a workflow, which allows me to write in VS Code in one window, and preview what I'm working on in a different browser window. I'm getting poor results with boolean operations in JSCAD, it's giving me non-manifold edges, which led me to this project.

So .... Is there something similar I can do with ManifoldCAD? VSCode integrates nicely with co-pilot and git, which is very helpful for developing larger projects. I understand that I can do this locally with Jupyter, but I'm primarily a Typescript user -- and at any rate Jupyter doesn't allow me to access co-pilot either.

**Comment by elalish** [comment]
Funny you should mention this: I was just suggesting to @pca006132 that we should author a VSCode extension that uses largely the same code base as ManifoldCAD.org. This would have several benefits beyond just the workflows you mentioned: nice filesystem support for importing and exporting meshes and libraries (a la OpenSCAD modules), auto-formatting and other editor niceties, and discoverability (via their store). 

Time being limited, I would love a collaborator. I don't suppose you know anything about authoring a VSCode extension by chance? At least it's all TS.

**Reply by tonyhschu** [reply]
I don't, unfortunately, but I am willing to look into it. From a brief glance, most of the relevant code is either in:
* worker.ts: https://github.com/elalish/manifold/blob/3b8282e1d5cd3d6f801432e4140e9b40f41ecbf6/bindings/wasm/examples/worker.ts ...or,
* three.html: https://github.com/elalish/manifold/blob/3b8282e1d5cd3d6f801432e4140e9b40f41ecbf6/bindings/wasm/examples/three.html

Is that right? 

Separately, it's not clear to me how VSCode extensions might load a WASM module, but at least it looks like it is possible: https://john-millikin.com/extending-vscode-with-webassembly

**Reply by elalish** [reply]
Yes, for worker.ts. However, the HTML is https://github.com/elalish/manifold/blob/3b8282e1d5cd3d6f801432e4140e9b40f41ecbf6/bindings/wasm/examples/index.html; three.html is an unrelated example. You'll also want editor.js: https://github.com/elalish/manifold/blob/3b8282e1d5cd3d6f801432e4140e9b40f41ecbf6/bindings/wasm/examples/editor.js, though we probably won't need all of it - the Monaco stuff can be factored out to a file that only ManifoldCAD.org uses, since VSCode already has the editor. 

The Monaco stuff, especially where we inject our TS library declarations, is the main reason that file is still JS - once that's removed it should be a lot easier to convert it to TS. Have you managed to get Vite running to serve a local dev instance of ManifoldCAD.org? Once you have that working it should be pretty easy to poke around and see how it's put together.

**Reply by tonyhschu** [reply]
Sorry I haven't had a lot of time to look into this. I'm not very familiar with C or WASM, so I'm getting stuck here:
`emcmake cmake -DCMAKE_BUILD_TYPE=Debug .. && emmake make` in this [README](https://github.com/elalish/manifold/blob/master/bindings/wasm/examples/README.md)

What's emmake? Is that just a typo of emcmake? Assuming that it is ... the next error I ran into looks like this
```
emcmake: error: 'make -DCMAKE_TOOLCHAIN_FILE=/opt/homebrew/Cellar/emscripten/3.1.51/libexec/cmake/Modules/Platform/Emscripten.cmake -DCMAKE_CROSSCOMPILING_EMULATOR=/opt/homebrew/opt/node/bin/node' failed (returned 2)
```

I'm on a M2 Mac. I ran the `cmake` command in the root readme without issue. I have done the `brew install nodejs` and `brew install emscripten` as well.

**Reply by elalish** [reply]
Not a typo - normally you run cmake to create the make files, then make to build the executable. The emscripten versions just put "em" in front of both commands. Does it work as written?

**Reply by pca006132** [reply, marked as answer]
I think for the vscode extension, it is fine to open another repo and use npm, so there is no need to build manifold by yourself.

**Reply by elalish** [reply]
Oh yeah, good call.

**Reply by tonyhschu** [reply]
Ok sounds good. I'll keep digging on the vscode extension front.

**Reply by tonyhschu** [reply]
Hey @elalish & @pca006132,

I dug into writing a VSCode extension a bunch - it felt like a lot of overhead. I went back and tried the `npx` command approach. Here's a very barebones prototype: https://github.com/tonyhschu/manifold-live

I still don't thoroughly understand how everything in `worker.ts` works, but I pulled together enough bits and pieces to have the following:
1. A nodejs server that serves a static HTML file
2. The server listens for changes to a target .ts file, and when a change occurs...
3. Imports the .ts file as a module, and gets the `manifold` object that results.
4. Writes a .glb file locally
5. The HTML, noticing a new timestamp on the server, refreshes itself and loads the new .glb.

It does roughly what I had set out to do, i.e. be able to write code on one screen, and see the result in the browser in a live-reload like manner. I would love guidance regarding the following:

1. How to wire up the code from `worker.ts` properly so that the animation, etc. also work.
2. Any advice on the module importing side -- I'm using `ts-import` which seems ... ok?

**Reply by elalish** [reply]
If you're using an unmodified `worker.ts`, you probably have animation export working fine. I'd guess you just need to grab the animation controls from `editor.js` and hook them up with your `index.html`. 

I like the sound of importing your live code as a TS module and running it. This will be key to allowing ManifoldCAD to use libraries, like OpenSCAD modules. I'll be interested to see how that works.

Can you give me any details of what you learned of VSCode extensions? It's unfortunate if there's a bunch of overhead, but I think that's the direction that makes more sense from a packaging and usability perspective.

**Reply by tonyhschu** [reply]
Regarding the VSCode Extension bits... I looked at [vscode-livepreview](https://github.com/microsoft/vscode-livepreview) (a microsoft-developed VSCode extension) as inspiration. It has two main parts:
- A Web View for previewing the content
- A Server for serving the content

For the web view there's VSCode's Web View API: https://code.visualstudio.com/api/extension-guides/webview which seemed relatively straight forward. It's the bit around figuring out how to run a web server in VSCode that seem complicated. [Scanning this serverGrouping.ts file in the vscode-livepreview plug-in](https://github.com/microsoft/vscode-livepreview/blob/main/src/server/serverGrouping.ts#L94), there's a lot of code that's just managing references and various set up/clean up type work for VSCode. Maybe a simple plug-in for Manifold doesn't need all that, but I didn't really want to find out when writing a NPX command involved only starting an ExpressJS server in the terminal (which I already knew how to do.)

That said, whatever server code I end up writing can probably be ported to a VSCode extension.

Also - I'm realizing as I write this that you probably want an extension that can deal with both Javascript code and Python code -- what I've done so far definitely does not do that. So maybe you do need a VSCode extension after all.

**Reply by pca006132** [reply]
I don't think we really need a web server. We can do [message passing](https://code.visualstudio.com/api/extension-guides/webview#passing-messages-from-an-extension-to-a-webview) from vscode extension to the webview, and there are APIs to listen to content changes + getting the code from the *unsaved* editor.

**Reply by pca006132** [reply]
regarding python, I don't think we need to consider that right now, it is easy enough to use the jupyter notebook, and we can always add python functionality to the extension later.

**Reply by elalish** [reply]
Thanks - but do we actually need a server? I agree that their Web View API looks great, but can't we accomplish everything we need through message passing? After all, even ManifoldCAD.org is just a static webpage - no server backend at all, and all communication with the worker is also through message passing. 

I'll probably hold off on Python for now - that already has a pretty solid framework just with Colab. Since VSCode is built in TS on Chromium, I think working in TS/JS will be a lot easier.

**Reply by elalish** [reply]
Ha, I should have refreshed - looks like we're on the same page, @pca006132. 😄

**Reply by danmarshall** [reply]
Hello all, and great to "see" you again Emmet! @elalish It's been a long time since the 3D Builder days in building 115 🤓🍻. I just found out about Manifold - congrats, and thanks! (I'll just keep following you from now on 😆)

Anyways as the author of [jscad-now](https://github.com/danmarshall/jscad-now) I thought I'd chime in with my lessons learned. 

> I like the sound of importing your live code as a TS module and running it. This will be key to allowing ManifoldCAD to use libraries, like OpenSCAD modules. I'll be interested to see how that works.

The idea of modules (dependencies) is key for me too. I suppose Python people will use pip and JavaScript people will use npm. Pleeease nobody create your own module system. 

Thankfully for JS this dependency business is all figured out now with modern bundlers.  If I were to start from scratch, I would use an off-the-shelf bundler/dev server like [Parcel](//[parceljs.org](https://parceljs.org/)) or [Vite](//vitejs.dev/). 

@tonyhschu your implementation has the same concepts (using node as a dev server, and a file watcher). Jscad-now did the same but it used Budo (ala Browserify but that is now old technology). I'd like to take a look at your manifold-live project when I get a chance.

**Reply by elalish** [reply]
Hey, great to see you again too, @danmarshall! Man, that was back before I'd ever written a line of C++. Agreed about not creating a new module system, and we're already using Vite for our internal development - it's quite nice!

That makes a lot of sense though - use Vite to bundle all of the user code, including dependencies, and send that to our worker to process. I think ideally I'd like to keep my own worker (with Manifold itself) separate, so we don't have to build and load that much larger piece of code over and over (plus that's how it works now). 

Are you still at M$? Do you have any familiarity with VSCode extensions? I'd really like to have cleaner file system support than is possible through a browser. Definitely looking for collaborators if anyone's interested!

**Reply by danmarshall** [reply]
Yup I'm still here 🦖, and *shameless plug* I maintain the [VSCode extension for SandDance](https://marketplace.visualstudio.com/items?itemName=msrvida.vscode-sanddance) . Agree on keeping your worker sovereign. If the Vite-based preview server is built, the VSCode extension can just be a lightweight wrapper around that.

**Reply by elalish** [reply]
Well good for you, Satya's done wonders for those stock units! SandDance looks super cool, I'm going to have to try that out. 

I'm hoping a bunch of [this code](https://github.com/elalish/manifold/blob/master/bindings/wasm/examples/editor.js#L289) from ManifoldCAD.org can be simplified in a VSCode extension version, while getting more functionality. You can see the Vite config I use for developing ManifoldCAD, but I'm less sure how to use it to bundle user-space code on the fly. If you want to sketch out some basic architecture diagram for how this might work, it would probably save me a lot of headaches.

**Reply by danmarshall** [reply]
I think I'd start with the most basic vanilla example. Is this as bare bones as it gets: https://manifoldcad.org/three.html ?

**Reply by hrgdavor** [reply]
@elalish in jscadui jscad.app, the worker intercepts require while maintaining information on the relative path so no bundling is needed. It converts esm js/ts code to cjs and executes it.

it can run remote code like this example:
https://jscad.app/#https://raw.githubusercontent.com/jscad/OpenJSCAD.org/master/packages/examples/import/STLImport/index.js

for developing projects in a local folder, I use drag and drop that gives file access in chrome and it can watch for changes. Only changed modules are reloaded. It is not perfect, but works rather well. 

That is also one way to go about it

**Reply by elalish** [reply]
@danmarshall Sort of, but that doesn't make use of our worker that runs scripts. Maybe this is a better place to start? https://manifoldcad.org/#Intro

I had been vaguely envisioning something like what @hrgdavor said, but perhaps Vite would simplify that? I'm not quite sure of the pros and cons.

**Reply by hrgdavor** [reply]
@elalish I have not weighted pros and cons much, old jscad worker was similar but limited. And with this I feel I have much more freedom, in some distant future to attach line numbers to generated geometries maybe.... dunno, It just feels rigth form at least.

I would like to create interactive tutorials (likely sometime when I retire in 20/30 yrs :D :D :D )

I did some draft samples 
https://3d.hrg.hr/jscad/three/loop/loop.html#1 with old regl renderer and the slower original jscad worker.

I have much more ideas than time :D ... still some of those explorations bring some useful things along the way (some that were not part of the end goal)

**Reply by danmarshall** [reply]
I think we're all part-timers here ⌛, which is why I emphasize leveraging common off-the-shelf building blocks. So imho, the top optimization is developer time and minimal code. 

Module resolution is a hairy problem, e.g. the use of `import`(es6 modules) vs `require` (commonjs). Thousands of human hours were spent to solve this issue. See [Rollup](https://rollupjs.org/introduction/#compatibility), [Parcel](https://parceljs.org/features/dependency-resolution/), [Vite](https://github.com/vite-plugin/vite-plugin-commonjs?tab=readme-ov-file#cases), [Observable import](https://observablehq.com/@observablehq/import), [Observable require debugger ⁉️](https://observablehq.com/documentation/debugging/require-stubborn-modules). We used to call it DLL Hell https://en.wikipedia.org/wiki/Dependency_hell. 

So that's why I promote a bundler.

**Reply by elalish** [reply]
Yeah, very good point. Okay, I guess I just need to figure out how to bring up Vite programmatically instead of on the command line. Actually, maybe I can even do it on the command line in a VSCode app? It has a terminal after all...

**Reply by hrgdavor** [reply]
> See Rollup, Parcel, Vite, Observable import, Observable require debugger 

thanks for the links, bundlers are always an option, but I still want to explore require exec  to be able to do more directly in the browser.
I do not have to solve dependency hell, but I think I can have sth useful that does not support dependencies from hell, but requires using deps that play nicely.

**Reply by danmarshall** [reply]
@elalish I was taking a look at it and came to the same conclusion that @tonyhschu did - much of the magic is in worker.ts - which makes me think worker.ts (or a great subset of it) should be compiled and published in the [manifold-3d npm package](https://www.npmjs.com/package/manifold-3d) as another module (that can still be invoked in a web worker). So that worker.ts does not need to be copied in the previewer project.

Specifically, this function `f` which takes code as a string: https://github.com/elalish/manifold/blob/c03a5ac3adbb002ef29646648adbbebba179000d/bindings/wasm/examples/worker.ts#L329-L333
ought to be slightly refactored and exposed as an `export`, referenceable by the preview server. The final piece of the puzzle would be a small shim file (lets call it shim.ts) which glues the worker export to customer's target .ts file.
```ts
\\shim.ts - written when previewer is invoked from command line
import f from 'manifold-3d/worker'
import * as customerCode from 'customertarget.ts' //populated by the command line parameter
f(customerCode)
```
Something like that.

**Reply by hrgdavor** [reply]
I had issue with this aproach, as it shifts line numbers in stack traces, is this shifting ignored, or there is a neat workaround to get  correct line numbers in exceptions ?

**Reply by pca006132** [reply]
https://stackoverflow.com/questions/1901012/javascript-exception-handling-displaying-the-line-number

**Reply by hrgdavor** [reply]
@pca006132 that is parsing after stack trace line numbers are messed up, and will still show wrong line numbers in console. That would be a last resort.

I do have sth that should work nicely, after trying different things

https://github.com/hrgdavor/jscadui/blob/fbff309ef1000602543f34f7c27143af0d3ce8b9/packages/require/src/require.js#L24
(I prefer making sure line numbers are correct when browser reports the err, with func above, or if it will fail, I may look into injecting sourcemap)

**Reply by elalish** [reply]
@danmarshall Agreed we should publish our worker to npm, and our gltf-io stuff too, so they can be more easily reused. What is the best way to package them? Should they be part of manifold-3d? Separate packages? Some kind of monorepo? I feel like I should understand npm better by now...

@hrgdavor Thanks, that's a nice little technique for stack trace line numbers - I hadn't realized that was a problem with `new Function()`.

**Reply by danmarshall** [reply]
I think added to the `manifiold-3d` npm package, as `worker.js` and `gltf-io.js`, as esm modules.

**Reply by hrgdavor** [reply]
> Thanks, that's a nice little technique for stack trace line numbers

Np, just added one small detail when calling the `runModule`
```
runModule(_require, exports, module, source + '\n//# sourceURL=' + url)
```
it gives also the proper file name 
![image](https://github.com/elalish/manifold/assets/2480762/14afe5df-aeb6-445d-8f2d-a9eebbd06f34)

unlike without the added `/# sourceURL`
![image](https://github.com/elalish/manifold/assets/2480762/d7422533-e0d0-4c47-b337-883620281769)
